### PR TITLE
Add Moderate Retry Mechanism

### PR DESCRIPTION
**Pull request by AradCarmi**
## Related Issues
fixes: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-7235)

## Description
Added moderate retry mechanism to using the `Oproxy` flow.

### CODE REVIEW

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py [hunk 1]
```diff
@@ -975,11 +978,17 @@ def _oproxy_authorize(self, resource: str = '', scope: str | None = None) -> tup
...
+        should_delay_request()
         oproxy_response = self._oproxy_authorize_build_request(headers, content, scope, resource)
 
         if not oproxy_response.ok:
+            next_request_time = calculate_next_request_time()
+            set_retry_mechanism_arguments(next_request_time=next_request_time)
             self._raise_authentication_error(oproxy_response)
 
+        # In case of success, reset the retry mechanism arguments.
+        set_retry_mechanism_arguments()
```
Comment (dantavori) [inline Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py:991]: let's do this in a single call to `get_integration_context()` and `set_integration_context()` (each). the reason is to reduce the number of interactions with the server.
also consider using a decorator

**Review by dantavori — COMMENTED** [review]
nice, see notes

**Review by dantavori — APPROVED** [review]
nice, approving but see my only note - don't merge before resolving this.

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py [hunk 2]
```diff
@@ -1404,6 +1416,46 @@ def __init__(self, message):
...
+        Sets the next_request_time in the integration context.
+        This is an implication of the Moderate Retry Mechanism for the Oproxy requests.
+    """
+    next_counter = delay_request_counter + 1
+    next_integration_context = {'next_request_time': next_request_time, 'delay_request_counter': next_counter}
+
+    # Should reset the context retry arguments.
+    if next_request_time == 0.0:
+        next_integration_context['delay_request_counter'] = 1
+    set_integration_context(next_integration_context)
```
[the remark this answers is not in the document]
Comment (AradCarmi) [reply Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py]: I updated the code not to override the context but I am not sure WDYM by other keys that we should remove from the context.

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py [hunk 3]
```diff
@@ -1426,18 +1427,20 @@ def calculate_next_request_time(delay_request_counter: int) -> float:
     return next_request_time.timestamp()
 
 
-def set_retry_mechanism_arguments(next_request_time: float = 0.0, delay_request_counter: int = 1):
+def set_retry_mechanism_arguments(context: dict, next_request_time: float = 0.0, delay_request_counter: int = 1):
     """
         Sets the next_request_time in the integration context.
         This is an implication of the Moderate Retry Mechanism for the Oproxy requests.
     """
+    context = context if context else {}
```
Comment (dantavori) [inline Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py:1435]: no need to commit if except for this you're ready, but:
```suggestion
    context = context or {}
```

**Review by dantavori — APPROVED** [review]
🚀

**Pull request merged**